### ISSUE TITLE
Message dropped after multiple "empty" subscribe/unsubscribes in PUB/SUB connection

### ISSUE THREAD

**Issue by pieleric** [zeromq/libzmq#2316]
We are encountering an issue with 0mq in our software. On a pub/sub connection with low HWM, if the subscriber does 2 SUBSCRIBE/UNSUBSCRIBE in a row without actually receiving any message, the next message(s) sent is dropped.

It does happen both with libzmq 4.1.4 and 4.1.6. This didn't happen with libzmq 2.1.11.

Here [zmqdrop.zip](https://github.com/zeromq/libzmq/files/719759/zmqdrop.zip) is an example program demonstrating the issue.
On one terminal, run `python server.py`, while on another run `python client.py`.
On the server side, the output is:
```
server is ready to listen
Sending one array
Sending one array
Sending one array
```
On the client side, the output (when the issue happens):
```
Going through empty cycles
Now asking for messages
Received new array (index = 1) after 2.06951 s
Received new array (index = 2) after 1.05825 s
```

While the expected client output is:
```
Going through empty cycles
Now asking for messages
Received new array (index = 0) after 1.01595 s
Received new array (index = 1) after 1.00752 s
Received new array (index = 2) after 1.00755 s
Stopping the server
Done
```

A workaround is to increase the HWM, but that's not handy for our case (and additional sub/unsub cycles can also trigger the issue). One other workaround I've found is in the server to explicitly change the HWM (back and forth) and unbind/rebind the connection before starting to send new messages.

**Comment by bluca** [comment]
Given you are using pyzmq I would suggest opening an issue on https://github.com/zeromq/pyzmq repository instead

**Comment by pieleric** [comment]
Sure I can.
Do you mean you think it's a bug in the python bindings, or that any issue with a test case in python should be opened in the pyzmq repo?

**Comment by bluca** [comment]
If you can reproduce a problem in c/c++ using only libzmq feel free to reopen and attach the code, but in general if you have code for a binding you can't really see where the problem is, and it's harder to debug given the layers of indirection

**Comment by pieleric** [comment]
As you suggested, I made a C version of the test case, and this hinted further on the issue. It happens only if you send the RCVHWM on the publisher side. Doing so doesn't seem documented and is probably not really meaningful anyway. Here is the example code: [zmqdropc.zip](https://github.com/zeromq/libzmq/files/724233/zmqdropc.zip).

This was caused by my misunderstanding of the python binding. I was modifying the .hwm attribute, which I thought was doing the "right thing" by changing the rcvhwm on subscribers and the sndhwm on publishers, but it actually does _both_. Explicitly changing just the sndhwm does lead to the correct behaviour.

So, I'd propose to close this report as it's not per se an issue, excepted if you consider that setting the RCVHWM on a socket configured for PUB should report an error or should have a documented behaviour.
